File: edd/Clase 5/Queue.hs
module Queue
    (Queue, emptyQ, isEmptyQ, queue, firstQ, dequeue)
    where

data Queue a = Q [a]
    deriving Show


emptyQ :: Queue a
-- Crea una cola vacía.
emptyQ = Q []

isEmptyQ :: Queue a -> Bool
-- Dada una cola indica si la cola está vacía.
isEmptyQ (Q []) = True
isEmptyQ _      = False

queue :: a -> Queue a -> Queue a
-- Dados un elemento y una cola, agrega ese elemento a la cola.
queue e (Q qs) = (Q (agregarAlFinal e qs))

agregarAlFinal :: a -> [a] -> [a]
agregarAlFinal a []     = [a]
agregarAlFinal a (x:xs) = x : (agregarAlFinal a xs)

firstQ :: Queue a -> a
-- Dada una cola devuelve el primer elemento de la cola.
firstQ (Q qs) = head qs

dequeue :: Queue a -> Queue a
-- Dada una cola la devuelve sin su primer elemento.
dequeue (Q qs) = (Q (tail qs))

File: edd/Clase 11/Queue.cpp
#include "Queue.h"
#include "iostream"

Queue emptyQ(){
    QueueSt* nuevoQ = new QueueSt;
    nuevoQ->cantidad = 0;
    nuevoQ->primero = NULL;
    nuevoQ->ultimo = NULL;
    return nuevoQ;
}
// Crea una lista vacía.
// Costo: O(1).

bool isEmptyQ(Queue q){
    return q->cantidad == 0;
}
// Indica si la lista está vacía.
// Costo: O(1).

int firstQ(Queue q){
    return q->primero->elem;
}
// Devuelve el primer elemento.
// Costo: O(1).

void Enqueue(int x, Queue q){
    NodoQ* nuevoNodo = new NodoQ;
    nuevoNodo->elem = x;
    nuevoNodo->siguiente = NULL;
    if(q->cantidad = 0){
        q->primero = nuevoNodo;
    }else{
        q->ultimo->siguiente = nuevoNodo;
    }
    q->ultimo = nuevoNodo;
    q->cantidad++;
}
// Agrega un elemento al final de la cola.
// Costo: O(1).

void Dequeue(Queue q){
    NodoQ* nodoASacar = q->primero;
    q->primero = nodoASacar->siguiente;
    delete nodoASacar;
}
// Quita el primer elemento de la cola.
// Costo: O(1).
// Precond: La cola debe de tener al menos un elemento

int lengthQ(Queue q){
    return q->cantidad;
}
// Devuelve la cantidad de elementos de la cola.
// Costo: O(1).

void MergeQ(Queue q1, Queue q2){
    while(!isEmptyQ(q2)){
        Enqueue(firstQ(q2), q1);
    }
    delete q2;
}
// Anexa q2 al final de q1, liberando la memoria inservible de q2 en el proceso.
// Nota: Si bien se libera memoria de q2, no necesariamente la de sus nodos.
// Costo: O(1).

void DestroyQ(Queue q){
    while(!isEmptyQ(q)){
        Dequeue(q);
    }
    delete q;
}
// Libera la memoria ocupada por la lista.
// Costo: O(n).